docs/design/2022-08-04-lazy-constraint-check.md

```diff
@@ -0,0 +1,138 @@
+INSERT INTO t1 VALUES (2); -- Skip acquiring the lock and return success.
+COMMIT; -- ERROR 1062 (23000): Duplicate entry '2' for key 'PRIMARY'
+```
+
+### Applicable Scenarios
+
+When talking about the success rate, we are based on the following assumption:
+
+**The user treats duplicated entry errors returned by DMLs as fatal errors which must abort the transaction.**
+
+A duplicated entry error only makes a single DML statement fail in a pessimistic transaction. The transaction does not abort immediately. If the user can handle the duplicated entry error in their application and continue the transaction, skipping pessimistic locks definitely changes the behavior and lowers the success rate.
+
+While under such an assumption, the general success rate won't change with the feature enabled in most cases.
+
+Suppose that transaction A wants to insert a new row with a unique key, or update a field with a unique key.
+
+1. If a duplicated record exists before the DML, when to check the constraint does not change the success rate. Currently, the duplicated entry error is raised when executing the DML. With lazy constraint check, the error is raised when committing the transaction. 
+
+2. If the duplicated record is written by transaction B between transaction A's executing the DML and committing, let's consider different cases:
+
+   1. If both transactions acquire the lock of the unique key, which is the current behavior, transaction B will be blocked until transaction A commits. After transaction A commits, the DML of transaction B will fail due to the duplicated entry.
+   2. If transaction A skips acquiring the lock while transaction B does not, transaction A will fail unless transaction B aborts. When transaction A prewrites, it will either meet the lock or the write record from transaction B. In either case, transaction A will fail to pass the constraint check.
+   3. If both transaction A and B skip acquiring the lock, it's similar to the case of an optimistic transaction. The transaction that commits first will succeed.
+
+   Therefore, although the time when the error is raised is different, only one of the transactions will fail.
+   
+In a special case, the success rate may drop due to the write conflict check:
+
+```sql
+/* init */ CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY);
+/* init */ INSERT INTO t1 VALUES (1);
+
+/* s1 */ BEGIN PESSIMISTIC;
+/* s2 */ DELETE FROM t1 WHERE id = 1;
+/* s1 */ INSERT INTO t1 VALUES (1);
+/* s1 */ COMMIT;
+```
+
+If we check constraints while acquiring the lock, `s1` will succeed because the row with `id=1` has been deleted when acquiring the lock. However, if we check constraints lazily, to ensure snapshot isolation, we have to do write a conflict check like in optimistic transactions in additional to constraint checks. In the case above, write conflict check will fail because the commit TS of `s2` is bigger than the start TS of `s1`.
```

Comment:
```suggestion
If we check constraints while acquiring the lock, `s1` will succeed because the row with `id=1` has been deleted when acquiring the lock. However, if we check constraints lazily, to ensure snapshot isolation, we have to do a write conflict check like in optimistic transactions in additional to constraint checks. In the case above, write conflict check will fail because the commit TS of `s2` is bigger than the start TS of `s1`.
```